lib/puppet/type/fullchain.rb

```diff
@@ -0,0 +1,48 @@
+Puppet::Type.newtype(:fullchain) do
+  @doc = 'Manage a full certificate chain in a single file. For a given
+  certificate file it will iteratively lookup and download the chain of
+  issuer certificates up until it find the self-signed root certificate
+  and store it in a single PEM formated file. It assumes, that all
+  certificates are accomodating the X.509v3 authorityInfoAccess extension
+  containing the caIssuer access URL. If a certificate in the chain does
+  not contain the proper information, the iteration will stop, log the
+  issue and write the incomplete chain.'
+
+  ensurable
+
+  newparam(:path, namevar: true) do
+    desc 'Path to the fullchain file to manage'
+    validate do |value|
+      unless Puppet::Util.absolute_path?(value)
+        raise Puppet::Error, _("File paths must be fully qualified, not '%{path}'") % { path: value }
+      end
+    end
+
+    munge do |value|
+      if value.start_with?('//') && File.basename(value) == '/'
+        # This is a UNC path pointing to a share, so don't add a trailing slash
+        File.expand_path(value)
+      else
+        File.join(File.split(File.expand_path(value)))
```

Comment:
What is the goal of splitting and joining again? Getting rid of double slashes?

Reply:
I don't get your point @rtib, on my system I get:

```ruby
File.expand_path('Projects/../bin')      #=> "/usr/home/romain/bin"
File.join(File.split('Projects/../bin')) #=> "Projects/../bin"
```

It is with Ruby 2.7 but I would be surprised to see a different behaviour with other versions.

```suggestion
        File.expand_path(value)
```

Reply:
Different Ruby versions handle trailing slashes differently. Don't remember the exact changes, but the expand_path of a 2.7.3 is not collapsing them properly:
```
irb(main):001:0> val='//tmp/text'
=> "//tmp/text"
irb(main):002:0> File.join(File.split(File.expand_path(val)))
=> "/tmp/text"
irb(main):003:0> File.expand_path(val)
=> "//tmp/text"
```

Reply:
I think this deserves some spec cases.